src/kudu/consensus/raft_consensus.cc

```diff
@@ -3297,6 +3335,280 @@ void RaftConsensus::SetLeaderDetectedCallback(LeaderDetectedCallback ldcb) {
+  if (request->has_last_idx_appended_to_leader()) {
+    downstream_request.set_last_idx_appended_to_leader(request->last_idx_appended_to_leader());
+  }
+
+  downstream_request.set_proxy_caller_uuid(peer_uuid());
+
+  string next_uuid = request->dest_uuid();
+  if (FLAGS_raft_enable_multi_hop_proxy_routing) {
+    RET_RESPOND_ERROR_NOT_OK(routing_table_->NextHop(peer_uuid(), request->dest_uuid(), &next_uuid));
+  }
+
+  if (request->dest_uuid() != next_uuid) {
+    // Multi-hop proxy request.
+    downstream_request.set_proxy_dest_uuid(next_uuid);
+    // Forward the existing PROXY_OP ops.
+    for (int i = 0; i < request->ops_size(); i++) {
+      *downstream_request.add_ops() = request->ops(i);
+    }
+    prevent_ops_deletion.cancel(); // The ops we copy here are not pre-allocated.
+  } else {
+
+    // Reconstitute proxied events from the local cache.
+    // If the cache does not have all events, we retry up until the specified
+    // retry timeout.
+    // TODO(mpercy): Switch this from polling to event-triggered.
+
+    vector<ReplicateRefPtr> messages;
+    do {
+
+      // We assume and enforce that a single request is composed of a range of ops.
+      int64_t first_op_index = -1;
+
+      int64_t max_batch_size = FLAGS_consensus_max_batch_size_bytes - request->ByteSizeLong();
+      for (int i = 0; i < request->ops_size(); i++) {
+        auto& msg = request->ops(i);
+        if (PREDICT_FALSE(msg.op_type() != PROXY_OP)) {
+          RET_RESPOND_ERROR_NOT_OK(Status::InvalidArgument(Substitute(
+              "proxy expected PROXY_OP but received opid {} of type {}",
+              OpIdToString(msg.id()),
+              OperationType_Name(msg.op_type()))));
+        }
+        if (i == 0) {
+          first_op_index = msg.id().index();
+        } else {
+          // TODO(mpercy): It would be nice not to require consecutive indexes in the batch.
+          // We should see if we can support it without a big perf penalty in IOPS.
+          if (PREDICT_FALSE(msg.id().index() != first_op_index + i)) {
+            RET_RESPOND_ERROR_NOT_OK(Status::InvalidArgument(Substitute(
+                "proxy requires consecutive indexes in batch, but received {} after index {}",
+                OpIdToString(msg.id()),
+                first_op_index + i - 1)));
+          }
+        }
+      }
+      // Now we know that all ops we are reconstituting are consecutive.
+
+      // TODO(mpercy): Check whether we have the event in our LogCache yet. If not,
+      // wait and retry, or subscribe to the event being available. Most likely we
+      // should try be simple / greedy and wait until we have all events in the
+      // cache? Or we could be aggressive and fill what we can?
+
+      OpId preceding_id;
+      // TODO(mpercy): Add an API for ReadOps() to take number of ops we want,
+      // instead of the max batch size.
+      if (request->ops_size() > 0) {
+        RET_RESPOND_ERROR_NOT_OK(queue_->log_cache()->ReadOps(first_op_index - 1,
```

Comment:
this first_op_index -1 is weird, but I guess I don't understand the intuition in this logcache API

Reply:
Yeah the LogCache API is kinda weird I agree. We might want to mess with it more but I didn’t want to take that on in my initial impl